accepted/2022/FunctionPointers.md

```diff
@@ -0,0 +1,769 @@
+
+When returning the calling conventions through `Type.GetFunctionPointerCallingConventions()`, any calling convention encoded using "CallKind" will return the appropriate CallConv* type, just like it was added through a modifier. This normalization avoids having to expose the "CallKind" in the API and makes it simpler for callers since there is just a single method to return the calling conventions no matter how they were encoded. The `Type.IsUnmanagedFunctionPointer` property was added since there are not CallConv* types for managed that would otherwise have helped determine this.
+
+Internally in the runtime, the "CallKind" information will no longer be tracked by the function pointer type. Currently it is tracked, but that will change for unmodified types to just detecting managed (via "default" and "vararg") to avoid inconsistent semantics across the older "CallKind" and newer "modopts" encoding formats.
+
+### _Off-topic thoughts on using custom modifiers to represent calling conventions_
+Using the CallConv* classes to represent calling conventions was added in V6 because new callling conventions needed to be added and would not fit within the 4 bits that were being used to encode the existing calling conventions based on ECMA-335. So the decision, arguably not the best, was made to encode new calling conventions using custom modifiers since that would not affect existing metadata readers or require a ECMA-335 change. However, there are alternative approaches which may or may not have been considered -- one such approach would be to change the encompassing byte to be a 32-bit compressed integer encoding instead which .NET/CLI uses extensively elsewhere. This compressed integer approach is possible since the encompassing byte did not use bit 7 (`0x80`) which is the flag for a compressed integer to switch from one byte to two bytes. This would require a ECMA-335 change which would break other metadata readers once two+ bytes are used.
+
+## Modified types
+Obtaining all custom modifiers including `ref\out\in` is useful for those who read metadata for varying reasons. The proposal to support that is to expose functionality that returns a "modified type" which, for function pointers, is a wrapper over the unmodified function pointer and overrides `GetFunctionPointerParameterTypes()` to return a potential "modified type" for each parameter. From those modified types, `GetRequiredCustomModifiers()` and `GetOptionalCustomModifiers()` can be used.
+
+A modified type always start with a "root" function pointer obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`.
```

Comment:
```suggestion
A modified type always start with a "root" type obtained from `FieldInfo.GetModifiedFieldType()`, `PropertyInfo.GetModifiedPropertyType()` and `ParameterInfo.GetModifiedParameterType()`.
```
I assume we are not going to special case the modified types for function pointers only. For example, the following will work:

```csharp
var initOnly = typeof(MyStruct).GetMethod("set_MyProperty").ReturnParameter;

Debug.Assert(initOnly.GetModifiedParameterType().GetRequiredCustomModifiers()[0] == typeof(IsExternalInit));


public struct MyStruct
{
    public int MyProperty { get; init; }
}
```

Reply:
> I assume we are not going to special case the modified types for function pointers only

This was actually a topic I wanted to discuss.

The `init` modifier belongs to the property setter declaration, and not to the `int` type, so this would work today:
```cs
Debug.Assert(initOnly.GetRequiredCustomModifiers()[0] == typeof(IsExternalInit));
```
but I wasn't necessarily assuming `GetModifiedParameterType()` would also include the modifiers (i.e. only function pointer types have embedded modifiers). However, since we are generalizing via "modified type" then it does make sense in the broader "modifier type" usages outside of function pointers. I'll update the design for this then.